### PR TITLE
lua: allow setting complex values in dynamicMetadata

### PR DESCRIPTION
**Pull request by vmg**
Description: This PR fixes a small TODO left by @dio: `Allow to set dynamic metadata using a table.` in the Lua API. Although the `get` function will convert any `Protobuf::Value` into its corresponding Lua type, the reverse `set` function would only allow plain strings to be passed as metadata.

The code turns out to be rather straightforward. The most controversial bit is the heuristic to detect whether a Lua table should be converted to a Protobuf `Struct` or `ListValue`. This heuristic is the one that most Lua JSON implementations use to decide whether a table is serialized as a JSON map or array.

Risk Level: Low
Testing: Unit tests for complex Lua structures & failure cases have been added
Docs Changes: The `lua_filter.rst` documentation has been updated to point out that other values besides strings can be passed as metadata.
Release Notes: Updated

cc @gorzell

Signed-off-by: Vicent Marti <vmg@strn.cat>

### CODE REVIEW

**Comment by dio** [comment]
@vmg Looks good, can you merge master?

File: docs/root/configuration/http/http_filters/lua_filter.rst [hunk 1]
```diff
@@ -504,7 +504,8 @@ set()
   dynamicMetadata:set(filterName, key, value)
 
 Sets key-value pair of a *filterName*'s metadata. *filterName* is a key specifying the target filter name,
-e.g. *envoy.lb*. The type of *key* and *value* is *string*.
+e.g. *envoy.lb*. The type of *key* is *string*. The type of *value* is any Lua type that can be mapped
+to a metadata value, including a Lua table with key/values.
```
Comment (dio) [inline docs/root/configuration/http/http_filters/lua_filter.rst]: I think we need an extended example here. Could you provide one?

And also we need to list the "acceptable" type for this: `MetadataMapHelper::loadValue` in here.

Thanks!

File: docs/root/intro/version_history.rst [hunk 2]
```diff
@@ -38,6 +38,7 @@ Version history
 * listeners: added :ref:`continue_on_listener_filters_timeout <envoy_api_field_Listener.continue_on_listener_filters_timeout>` to configure whether a listener will still create a connection when listener filters time out.
 * listeners: added :ref:`HTTP inspector listener filter <config_listener_filters_http_inspector>`.
 * lua: extended `httpCall()` and `respond()` APIs to accept headers with entry values that can be a string or table of strings.
+* lua: allow setting complex values in dynamicMetadata
```
Comment (dio) [inline docs/root/intro/version_history.rst]: nit: extended `dynamicMetadata():set()` to allow...

File: source/extensions/filters/common/lua/wrappers.cc [hunk 3]
```diff
@@ -87,6 +87,99 @@ void MetadataMapHelper::createTable(lua_State* state,
...
+  default:
+    luaL_error(state, "unexpected type '%s' in dynamicMetadata", lua_typename(state, type));
+  }
+
+  return value;
+}
+
+/**
+ * Returns the length of a Lua table if it's actually shaped like a List,
+ * i.e. if all the keys are consecutive number values. Otherwise, returns -1
```
Comment (dio) [inline source/extensions/filters/common/lua/wrappers.cc]: super duper nit: missing full-stop.

File: source/extensions/filters/common/lua/wrappers.cc [hunk 4]
```diff
@@ -87,6 +87,99 @@ void MetadataMapHelper::createTable(lua_State* state,
...
+}
+
+ProtobufWkt::Struct MetadataMapHelper::loadStruct(lua_State* state) {
+  ProtobufWkt::Struct struct_obj;
+
+  lua_pushnil(state);
+  while (lua_next(state, -2) != 0) {
+    int key_type = lua_type(state, -2);
+    if (key_type != LUA_TSTRING) {
+      luaL_error(state, "unexpected type %s in table key (only string keys are supported)",
```
Comment (dio) [inline source/extensions/filters/common/lua/wrappers.cc]: Nice! We should have this documented.

File: source/extensions/filters/http/lua/wrappers.cc [hunk 5]
```diff
@@ -156,11 +156,19 @@ int DynamicMetadataMapWrapper::luaSet(lua_State* state) {
...
   }
 
-  // TODO(dio): Allow to set dynamic metadata using a table.
   const char* filter_name = luaL_checkstring(state, 2);
   const char* key = luaL_checkstring(state, 3);
-  const char* value = luaL_checkstring(state, 4);
-  streamInfo().setDynamicMetadata(filter_name, MessageUtil::keyValueStruct(key, value));
+
+  // MetadataMapHelper::loadValue will convert the value on top of the Lua stack,
+  // so push a copy of the 3rd arg ("value") to the top
```
Comment (dio) [inline source/extensions/filters/http/lua/wrappers.cc]: sorry, nit. full-stop. And everywhere else.

File: test/extensions/filters/http/lua/wrappers_test.cc [hunk 6]
```diff
@@ -311,6 +311,81 @@ TEST_F(LuaStreamInfoWrapperTest, SetGetAndIterateDynamicMetadata) {
   wrapper.reset();
 }
 
+// Set, get complex key/values in stream info dynamic metadata.
```
Comment (dio) [inline test/extensions/filters/http/lua/wrappers_test.cc]: Mind to extend one in lua_filter_test and lua_integration_test? Most of the time takin a look at these two tests is the best way to learn on the usability of it.

**Comment by vmg** [comment]
@dio: fixed all your comments and merged master.

**Comment by vmg** [comment]
:bow: Thank you so much!

**Pull request merged**